R/utils.R

```diff
@@ -103,46 +103,72 @@ Warn = function(msg, ...) rlang::warn(break_str(msg, init = "Warning: "), ...)
 #' Assert that a sliding computation function takes enough args
 #'
 #' @param f Function; specifies a computation to slide over an `epi_df` or
-#'  `epi_archive` in `epi_slide` or `epix_slide`.
+#'   `epi_archive` in `epi_slide` or `epix_slide`.
+#' @param ... Dots that will be forwarded to `f` from the dots of `epi_slide` or
+#'   `epix_slide`.
 #'
 #' @importFrom rlang is_missing
 #' @importFrom purrr map_lgl
 #'
 #' @noRd
-assert_sufficient_f_args <- function(f) {
-  n_mandatory_f_args <- 2
-  args_f = formals(args(f))
-  arg_names = names(args_f)
-  if ("..." %in% arg_names) {
+assert_sufficient_f_args <- function(f, ...) {
+  mandatory_f_args_labels <- c("window data", "group key")
+  n_mandatory_f_args <- length(mandatory_f_args_labels)
+  args = formals(args(f))
+  args_names = names(args)
+  # Remove named arguments forwarded from `epi[x]_slide`'s `...`:
+  forwarded_dots_names = names(rlang::call_match(dots_expand = FALSE)[["..."]])
+  args_matched_in_dots =
+    # positional calling args will skip over args matched by named calling args
+    args_names %in% forwarded_dots_names &
+    # extreme edge case: `epi[x]_slide(<stuff>, dot = 1, `...` = 2)`
+    args_names != "..."
+  remaining_args = args[!args_matched_in_dots]
+  remaining_args_names = names(remaining_args)
+  # note that this doesn't include unnamed args forwarded through `...`.
+  dots_i <- which(remaining_args_names == "...") # integer(0) if no match
+  if (length(dots_i) != 0L) {
     # Keep all arg names before `...`
-    dots_i <- which(arg_names == "...")
-    arg_names_before_dots <- arg_names[seq_len(dots_i - 1)]
-
-    if (length(arg_names_before_dots) < n_mandatory_f_args) {
-      Warn(sprintf("`f` only takes %s positional arguments before the `...` args, but `epi[x]_slide` will call it with at least %s positional arguments; if `f` doesn't expect those arguments, it may produce confusing error messages", length(arg_names), n_mandatory_f_args),
-           class="epiprocess__assert_sufficient_f_args__f_needs_min_args_before_dots",
-          epiprocess__f = f,
-          epiprocess__arg_names_before_dots = arg_names_before_dots)
+    mandatory_args_mapped_names <- remaining_args_names[seq_len(dots_i - 1L)]
+
+    if (dots_i - 1L < n_mandatory_f_args) {
```

Comment:
suggestion (non-blocking): We use `dots_i - 1L` a number of times. Defining a variable to store it would make the code more readable.